packages/@aws-cdk/cdk/lib/app.ts

```diff
@@ -1,20 +1,98 @@
 import cxapi = require('@aws-cdk/cx-api');
-import fs = require('fs');
-import path = require('path');
 import { Stack } from './cloudformation/stack';
-import { IConstruct, MetadataEntry, PATH_SEP, Root } from './core/construct';
+import { Construct } from './core/construct';
+import { IConstruct, MetadataEntry, PATH_SEP } from './core/construct';
+import { Environment } from './environment';
+import { Program } from './program';
+import { validateAndThrow } from './util/validation';
 
 /**
- * Represents a CDK program.
+ * Properties for an App
  */
-export class App extends Root {
+export interface AppProps {
+  /**
+   * The AWS environment (account/region) where stacks in this app will be deployed.
+   *
+   * If not supplied, the `default-account` and `default-region` context parameters will be
+   * used. If they are undefined, it will not be possible to deploy the stack.
+   *
+   * @default Automatically determined
+   */
+  env?: Environment;
+
+  /**
+   * The CDK Program instance in which this app will be defined.
+   *
+   * You don't have to pass this, it exists for testing purposes. Only
+   * supply this parameter if you know what you are doing.
+   *
+   * @default Automatically created
+   */
+  program?: Program;
+}
+
+/**
+ * Used as construct name if no ID is supplied for the App
+ */
+const DEFAULT_APP_NAME = 'App';
+
+/**
+ * An instance of your App
+ */
+export class App extends Construct {
+  /**
+   * True if the given construct is a default-named App
+   */
+  public static isDefaultApp(construct: IConstruct) {
+    return App.isApp(construct) && construct.defaultAppName;
+  }
+
+  /**
+   * True if the given construct is an App object
+   */
+  public static isApp(construct: IConstruct): construct is App {
+    return (construct as any).defaultAppName !== undefined;
+  }
+
+  /**
+   * The environment in which this stack is deployed.
+   */
+  public readonly env: Environment;
+
+  /**
+   * Whether the App id was supplied
+   */
+  private defaultAppName: boolean;
+
+  /**
+   * Whether validate() was already run or not
+   */
+  private validated: boolean;
+
+  /**
+   * Whether prepare() was already run or not
+   */
+  private prepared: boolean;
+
   /**
    * Initializes a CDK application.
+   *
    * @param request Optional toolkit request (e.g. for tests)
    */
-  constructor() {
-    super();
-    this.loadContext();
+  constructor(id?: string, props: AppProps = {}) {
```

Comment:
Since we are breaking the construct signature anyway here, why not make ID and optional prop instead of an optional argument?

Reply:
Cf:

```
const app = new App('MyApp');

const app = new App({ id: 'MyApp' });
```

Even though we don't have the full construct signature, at least in this it's somewhat obvious that the ID passed is a regular construct ID, which will contribute to a construct path (a stack would have path `/MyApp/MyStack`). I don't find that as obvious in the second case. Plus, the second one reads more noisily. Just generally don't care for it much.